servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ProxyConnectConnectionFactoryFilter.java

```diff
@@ -100,28 +103,43 @@ public void userEventTriggered(final ChannelHandlerContext ctx, final Object evt
                         }
                     });
 
-                    DeferSslHandler deferSslHandler = channel.pipeline().get(DeferSslHandler.class);
+                    final DeferSslHandler deferSslHandler = channel.pipeline().get(DeferSslHandler.class);
                     if (deferSslHandler == null) {
-                        return response.payloadBodyAndTrailers().ignoreElements().concat(failed(
-                                new IllegalStateException("Failed to find a handler of type " +
-                                        DeferSslHandler.class + " in channel pipeline.")));
+                        return failed(new IllegalStateException("Failed to find a handler of type " +
+                                DeferSslHandler.class + " in channel pipeline."));
                     }
-
                     deferSslHandler.ready();
 
-                    // There is no need to apply offloading explicitly (despite completing `processor` on the
-                    // EventLoop) because `payloadBody()` will be offloaded according to the strategy for the
-                    // request.
-                    return response.payloadBodyAndTrailers().ignoreElements().concat(fromSource(processor));
-                } else {
-                    return response.payloadBodyAndTrailers().ignoreElements().concat(
-                            failed(new ProxyResponseException("Bad response from proxy CONNECT " + connectAddress,
-                                    response.status())));
-                }
-            }));
+                    return fromSource(processor);
+                    // Close recently created connection in case of any error or cancellation while it connects to proxy
+                 }).whenFinally(new TerminalSignalConsumer<C>() {
+                    @Override
+                    public void onSuccess(@Nullable final C result) {
+                        // noop
+                    }
+
+                    @Override
+                    public void onError(final Throwable ignore) {
+                        closeConnection(c);
+                    }
+
+                    @Override
+                    public void cancel() {
+                        closeConnection(c);
```

Comment:
It seems cancel path may leak connection in general if the connection got emitted from here but the source got cancelled afterwards. eg: any operator added after this proxy filter may not send the connection further if it sees a `cancel()`.

Reply:
Interesting observation. I was focused on cancel handling before `onSuccess`. For example, before the handshake completes: https://github.com/apple/servicetalk/pull/1002/files#diff-09ab1790ede42935bdec34cd0d96560cR234-R251

Will you suggest to use `whenCancel` operator and always close the connection on cancellation? Because close is idempotent it should not matter if this connection is already closed by someone else. Is there any use-case when the connection is still useful after cancel?

Reply:
This is a more broader problem and it can not be handled specifically in this class. As `cancel()`s can come after a success, we really do not know if the cancel has come due to the operation was cancelled by the user or due to some operators sending a cancel for the previous source when they move on to the next source (eg: `concat()`).

I commented here to highlight the problem but this PR does not have to be blocked for this change. What you have here is sufficient in isolation for this class.